### PR TITLE
Refactor store structure & add support for list

### PR DESCRIPTION
**Pull request by FieryCod**
**What is done?**

- refactor of our store structure (resources are added, logic is flattened now)
- basic category page
- composer

**What should be done next?**
- Fetching more articles from the category
- Categories as a resource
- Styles for the list in category page (they are not the same as layout says)
- Categories widget (Here is a question how should we present the widget on mobile?)

If you like the idea of composers I can write some docs which will describe the idea behind it.

**What is available in compose method?**

*before* - Acts before the calling of action (if *callActionWhen* it does not return the true then the action, after, fail are not called)
*success* - after successful action call
*fail* - when the action does not succeed
*always* - always called after everything

### CODE REVIEW

File: store/view/articles.js [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+  currentCategoryCode: state => state.currentCategoryCode,
+  categoryCodeList: state => state.currentCategoryCode === 'wszystkie' ? state.mainlist : state.byCategory[state.currentCategoryCode],
+  byCategory: state => state.byCategory
+}
+
+// Module state
+export const state = () => ({
+  totalItems: 0,
+  mainpage: [],
+  mainlist: [],
```
Comment (k911) [inline store/view/articles.js]: `main list` and `main page` are two words, so I think we should change that to `mainPage` and `mainList`

File: store/view/articles.js [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+  getMainlist: composer.compose({
+    before: [
+      compounds.doSth(({ ctx }) => ctx.commit(changeCurrCategory('wszystkie'))),
+      compounds.callActionWhen(({ ctx }) => _.isEmpty(ctx.getters.mainlist)),
+      compounds.doSth(({ ctx }) => ctx.commit(loading(true)))
+    ],
+    caller: callers.squashErrorsCall('#getMainlist'),
+    success: [
+      opuses.doSth(({ ctx, result }) => ctx.dispatch('addCodes', { path: 'mainlist', codes: result })),
+      opuses.doSth(({ ctx, result }) => ctx.dispatch('addCodes', { path: 'all', codes: result })),
```
Comment (k911) [inline store/view/articles.js]: Either `addCodes` have issue in logger string, which says: adding to 'all' and '${path}`
or you don't need this line

File: store/view/articles.js [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+}
+
+export const mutations = {
+  [ types.LOAD_ARTICLES ] (state, payload) {
+    state.loading = payload.loading
+  },
+  [ types.ADD_CODES ] (state, payload) {
+    if (payload.path) {
+      state[payload.path] = _.union(state[payload.path], payload.codes)
+    } else {
```
Comment (k911) [inline store/view/articles.js]: I don't really like this one (having two different behaviours), is there something we could do @FieryCod ? Like split to TWO mutations? or it would be to much impractical?

File: store/view/articles.js (hunk 2 above)
Comment (FieryCod) [reply store/view/articles.js]: Logger has wrong string thanks :+1:

File: store/resources.js [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+  articles: state => state.articles
+}
+
+// Module state
+export const state = () => ({
+  articles: {}
+})
+
+export const actions = {
+  injectResource ({ state, commit }, params) {
```
Comment (k911) [inline store/resources.js]: injectResource**s**, I think

File: store/view/articles.js [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+const loading = storeHelper.createMutationFn(types.LOAD_ARTICLES, 'loading')
+const codes = storeHelper.createMutationFn(types.ADD_CODES)
+
+// Module getters
+export const getters = {
+  loading: state => state.loading,
+  mainpage: state => state.mainpage,
+  mainlist: state => state.mainlist,
+  all: state => state.all,
+  currArticleCode: state => state.currArticleCode,
```
Comment (k911) [inline store/view/articles.js]: Same as bellow

File: store/view/articles.js (hunk 3 above)
Comment (FieryCod) [reply store/view/articles.js]: Well, I do not like it either. But I do not know whether having two mutations like
`ADD_CODES_CATEGORY` or `ADD_CODES_TO_STH` makes sense here. I don't know. @zrebcu411 could you propose some solution maybe?

File: store/view/articles.js [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+
+    // Inject articles and return their codes
+    return dispatch('resources/injectResource', {
+      path: 'articles',
+      data: await dispatch('prepareArticles', data)
+    }, { root: true })
+  }),
+  getMainlist: composer.compose({
+    before: [
+      compounds.doSth(({ ctx }) => ctx.commit(changeCurrCategory('wszystkie'))),
```
Comment (k911) [inline store/view/articles.js]: String 'wszystkie' should be as const in settings?

File: store/resources.js (hunk 4 above)
Comment (FieryCod) [reply store/resources.js]: Well, I was thinking about injectResources but then I came with an idea that this action injects only one type of resource per call with the unlimited number of items. Maybe I should change the action name then? What do you think @k911 ?

**Comment by k911** [comment]
Really good and I think hard work! I'll look into it once more time, when others do theirs reviews!

**Comment by FieryCod** [comment]
@k911 Thank you very much I'm really thankful for your appreciation.
I gonna fix the issues when everyone will take a voice.

File: components/article/ArticleCardList.vue [hunk 7]
```diff
@@ -32,103 +33,71 @@ export default {
...
+@import "assets/scss/_imports";
 
 .article-card-list {
   display: flex;
   flex-wrap: wrap;
 
-  .article-card {
-    margin: 0 $default-gutters-width $default-gutters-width 0;
+  @media (max-width: $screen-sm) {
+    padding: 0 10px;
```
Comment (zrebcu411) [inline components/article/ArticleCardList.vue]: why did you change a variable to px again? It's important because I decided some time ago to change space between components from 20px to 15px. I simply changed gutters variable and said 'Let's magic happen' but nothing happens.. :(

File: helpers/template.js [hunk 8]
```diff
@@ -0,0 +1,73 @@
+import _ from 'lodash'
+
+/**
+ * Creates list of comma separated values
+ * @param {Array<String>} arr
+ * @returns {String}
+ */
+function strArrToStrList (arr) {
+  return _.reduce(arr, (str, item) => `${str}, ${item}`)
```
Comment (zrebcu411) [inline helpers/template.js]: totally useless helper, just use: `arr.join(', ')`

File: helpers/template.js [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+
+/**
+ * Returns string with dots if
+ * length of a string is too big
+ * @param {String} str
+ * @param {Number} [length=3]
+ * @returns {String}
+ */
+function trimString (str, length = 3) {
+  return (length - 3) >= str.length ? str : `${str.slice(0, length - 3)}...`
```
Comment (zrebcu411) [inline helpers/template.js]: I prefer to rename this function to `ellipsis(str, length = 3)`, because this is not only trimming but making 'wielokropek' too

File: store/view/articles.js [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+
+    // Inject articles and return their codes
+    return dispatch('resources/injectResource', {
+      path: 'articles',
+      data: await dispatch('prepareArticles', data)
+    }, { root: true })
+  }),
+  prepareArticles (ctx, data) {
+    const articlesArr = commonHelper.pickItemsProps(data['hydra:member'], settings.props, true, settings.datePicker)
+    return _.keyBy(articlesArr, article => article.code)
```
Comment (zrebcu411) [inline store/view/articles.js]: `_.keyBy(articlesArr, 'code')`

File: store/view/articles.js (hunk 3 above)
Comment (zrebcu411) [reply store/view/articles.js]: I think it's ok :)

File: components/article/ArticleCard.vue [hunk 11]
```diff
@@ -2,35 +2,57 @@
...
-      <router-link :to="{ name: 'articles-slug', params: { slug } }" class="article-card__title-link">
+
+      <!-- Article title -->
+      <router-link :to="{ path: code }" class="article-card__title-link">
         <h3 ref="articleTitle" class="article-card__title" itemprop="headline"> {{ trimString(title, 75) }} </h3>
       </router-link>
+
+      <!-- Article tags -->
+      <ul class="article-card__tags-wrapper">
+        <li class="article-card__single-tag" v-for="(tag, index) in tags" :key="index"> {{ tag.name }} </li>
```
Comment (zrebcu411) [inline components/article/ArticleCard.vue]: Please fix text wrapping

File: store/resources.js (hunk 4 above)
Comment (k911) [reply store/resources.js]: @FieryCod Definitely we should allow to inject only one type of resource at a time, but still I think that injectResources is more descriptive but whatever you decide I'm ok with it

**Comment by FieryCod** [comment]
@maksrojek @LaYatiX Any additional comments?

File: store/resources.js (hunk 4 above)
Comment (FieryCod) [reply store/resources.js]: @k911 I decided to stick to injectResource. I hope that you don't mind.

**Comment by FieryCod** [comment]
@knit-pk/web May we merge this?

File: components/commons/KnitNavbar.vue [hunk 12]
```diff
@@ -39,7 +39,7 @@ export default {
         },
         {
           text: 'Artykuły',
-          path: '/articles'
+          path: '/wszystkie'
```
Comment (k911) [inline components/commons/KnitNavbar.vue]: I don't really like Polish name in path R W sure we wanna do this?

File: pages/_categoryCode/index.vue [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+    ...mapGetters({
+      categoryCodeList: 'view/articles/categoryCodeList',
+      isMobile: 'general/isMobile'
+    }),
+    list () {
+      return _.map(this.categoryCodeList, code => this.$store.getters[ 'resources/articles' ][code])
+    }
+  },
+  fetch ({ store, params }) {
+    if (params.categoryCode === 'wszystkie') {
```
Comment (k911) [inline pages/_categoryCode/index.vue]: 'wszystkie' should be const

File: components/commons/KnitNavbar.vue (hunk 12 above)
Comment (zrebcu411) [reply components/commons/KnitNavbar.vue]: I totally agree with this. `Wszystkie` sounds really, really weird.

Comment (FieryCod) [reply components/commons/KnitNavbar.vue]: So how would you name all articles in path?
`/wszystkie` seems good to me

Comment (FieryCod) [reply components/commons/KnitNavbar.vue]: `wszystkie` simulates the category while articles does not

File: pages/_categoryCode/index.vue (hunk 13 above)
Comment (FieryCod) [reply pages/_categoryCode/index.vue]: It's used only once I cannot get why should I create const for it. Should import the name from settings or what?

File: components/commons/KnitNavbar.vue (hunk 12 above)
Comment (zrebcu411) [reply components/commons/KnitNavbar.vue]: The perfect solution for me would be `/articles` for all articles and '/articles/:category' for a specific category. Is it possible? If not, i suggest `/articles/all`

File: pages/_categoryCode/index.vue (hunk 13 above)
Comment (k911) [reply pages/_categoryCode/index.vue]: We using it three times in whole codebase, and and usage is already a declared const, in my opinion it should be strict const with uppercased name in seperate 'consts' file where we could add more like this: `const CATEGORY_CODE_ALL = 'wszystkie'`

File: components/commons/KnitNavbar.vue (hunk 12 above)
Comment (k911) [reply components/commons/KnitNavbar.vue]: @zrebcu411 for me that's perfect

Comment (FieryCod) [reply components/commons/KnitNavbar.vue]: I would disagree with that. We should prepare friendly urls because of search engines. From my experience with content products the best option is always present the url like `/categoryCode/slug/shortid`

File: pages/_categoryCode/index.vue (hunk 13 above)
Comment (FieryCod) [reply pages/_categoryCode/index.vue]: Where then the const file should be placed?

Comment (k911) [reply pages/_categoryCode/index.vue]: 🗡  models/Category/CategoryInterface

File: components/commons/KnitNavbar.vue (hunk 12 above)
Comment (k911) [reply components/commons/KnitNavbar.vue]: no problem but we'll have `projects` right?

Comment (zrebcu411) [reply components/commons/KnitNavbar.vue]: @k911 supporting two langs in a content targeted site is not possible I think. Whole content will be created by KNIT members, and we cannot tell them to write articles, offers etc. in two languages. On the other hand, if we'll be translating only specific parts of site e.g. menu, the site will be not consistent.

File: store/resources.js (hunk 4 above)
Comment (k911) [reply store/resources.js]: @FieryCod Yeah, no problem 👍

File: components/commons/KnitNavbar.vue (hunk 12 above)
Comment (FieryCod) [reply components/commons/KnitNavbar.vue]: As far as I know we do not want to create multi language site. Imagine how hard we would have to work to mantain multilanguage content site. 

Look how others do content services.
https://moz.com/learn/seo/url
https://wiadomosci.wp.pl/kaczynski-doszlismy-do-celu-kolejne-marsze-beda-raz-na-rok-6239808818247809v
http://www.blikk.hu/aktualis/politika/igy-vezetett-vona-gabor-utja-a-lemondasig/29q9w97
https://www.theguardian.com/world/europe-news
https://www.washingtonpost.com/business/economy/for-mark-zuckerberg-protecting-facebook-means-stepping-into-the-spotlight/2018/04/09/31518554-3c0a-11e8-a7d1-e4efec6389f0_story.html?hpid=hp_hp-top-table-main_zuckerberg-240p%3Ahomepage%2Fstory&noredirect=on

Imagine what users would feel when they see url like
 `baseurl/articles/category/slug/id`

It is possible that the slug would not be visible which is not something we would like to have.

Comment (FieryCod) [reply components/commons/KnitNavbar.vue]: Events, projects etc are not the main purpose of our site. The main purpose is of course presenting the articles. Others might have path startig from events, projects and so on.

Comment (zrebcu411) [reply components/commons/KnitNavbar.vue]: @FieryCod all examples you presented relate to the single resource, not list

Comment (k911) [reply components/commons/KnitNavbar.vue]: There sereval options to resolve that issue (e.g. like wikipedia, we could mark an article in a future if it is translated, so that's no problem)
i18n != supporting one optional language
All I'm saying is that we should not block possibility in future, or result in complicated refactor if we decide to.

File: pages/_categoryCode/index.vue (hunk 13 above)
Comment (FieryCod) [reply pages/_categoryCode/index.vue]: That was not helpful brother xD

File: components/commons/KnitNavbar.vue (hunk 12 above)
Comment (FieryCod) [reply components/commons/KnitNavbar.vue]: @zrebcu411 remove slug and then everything is list

File: pages/_categoryCode/index.vue (hunk 13 above)
Comment (k911) [reply pages/_categoryCode/index.vue]: I would keep this alongside logic connected with category, but of course we don't have this yet

File: components/commons/KnitNavbar.vue (hunk 12 above)
Comment (FieryCod) [reply components/commons/KnitNavbar.vue]: @k911 We will not support ever multilanguage website. Services with articles are in one main language. So we might switch to english in whole site or just use Polish.

File: pages/_categoryCode/index.vue (hunk 13 above)
Comment (FieryCod) [reply pages/_categoryCode/index.vue]: It's better idea i like it. What do you think @zrebcu?

File: components/commons/KnitNavbar.vue (hunk 12 above)
Comment (k911) [reply components/commons/KnitNavbar.vue]: @FieryCod @zrebcu411 Have you ever been to wikipedia? Why we should block opportunity to write article in english? What if somebody wrote article even in chineese? 😄

Comment (k911) [reply components/commons/KnitNavbar.vue]: And remeber that KNIT is not all about news content, it is supposed to be a platform, however news content is our primary goal

File: pages/_categoryCode/index.vue (hunk 13 above)
Comment (zrebcu411) [reply pages/_categoryCode/index.vue]: I think it's not config. Maybe new directory `constants/common.js`?

File: components/commons/KnitNavbar.vue (hunk 12 above)
Comment (FieryCod) [reply components/commons/KnitNavbar.vue]: Wikipedia does not have articles I believe and it's whole idea is to present bare informations.

File: pages/_categoryCode/index.vue (hunk 13 above)
Comment (FieryCod) [reply pages/_categoryCode/index.vue]: Then the constants folder will be introduced.

File: components/commons/KnitNavbar.vue (hunk 12 above)
Comment (zrebcu411) [reply components/commons/KnitNavbar.vue]: Wikipedia is a specific site, it is a worldwide encyclopedia. Our site aims to connect specific group of people

Comment (FieryCod) [reply components/commons/KnitNavbar.vue]: Yup totally agree with you @zrebcu411

File: pages/_categoryCode/index.vue (hunk 13 above)
Comment (k911) [reply pages/_categoryCode/index.vue]: soon we'll end up in 20+ top level directories, I don't think it's good way to organize project..
Imagine we'll have some specific functions for articles, we would create 'helpers/aricleHelper'? etc?
Why not create single directory for domain logic? Like `domain/articles` or similar? Then there could be functions, objects, consts related with `articles`

File: components/commons/KnitNavbar.vue (hunk 12 above)
Comment (zrebcu411) [reply components/commons/KnitNavbar.vue]: @FieryCod show me an example of a site with a list of ALL articles and separate smaller sublists. In all sites, a list of all articles is on mainpage

Comment (k911) [reply components/commons/KnitNavbar.vue]: We lost the point, do you really like polish name in url? @FieryCod @zrebcu411?

Comment (FieryCod) [reply components/commons/KnitNavbar.vue]: @zrebcu411 That's not true! On mainpage they got only selected subset of articles. That's true that they does not have special category like `wszystkie` simply because they got so many articles that they do not need it.
